hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/observer/ObservableRepository.java

```diff
@@ -0,0 +1,226 @@
+import com.hazelcast.jet.Observable;
+import com.hazelcast.jet.config.JetConfig;
+import com.hazelcast.jet.core.JetProperties;
+import com.hazelcast.jet.core.ProcessorMetaSupplier;
+import com.hazelcast.jet.datamodel.Tuple2;
+import com.hazelcast.spi.properties.HazelcastProperties;
+import com.hazelcast.topic.ITopic;
+import com.hazelcast.topic.Message;
+import com.hazelcast.topic.ReliableMessageListener;
+
+import javax.annotation.Nonnull;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.UUID;
+import java.util.function.Consumer;
+import java.util.function.LongSupplier;
+
+import static com.hazelcast.jet.impl.JobRepository.INTERNAL_JET_OBJECTS_PREFIX;
+import static java.lang.Math.min;
+
+public class ObservableRepository {
+
+    /**
+     * Constant ID to be used as a {@link ProcessorMetaSupplier#getTags()
+     * PMS tag key} for specifying when a PMS owns an {@link Observable} (ie.
+     * is the entity populating the {@link Observable} with data).
+     */
+    public static final String OWNED_OBSERVABLE = "owned_observable";
+
+    /**
+     * Prefix of all topic names used to back {@link Observable} implementations,
+     * necessary so that such topics can't clash with regular topics used
+     * for other purposes.
+     */
+    private static final String JET_OBSERVABLE_NAME_PREFIX = INTERNAL_JET_OBJECTS_PREFIX + "observable.";
+
+    private static final String COMPLETED_OBSERVABLES_LIST_NAME = INTERNAL_JET_OBJECTS_PREFIX + "completedObservables";
+    private static final int MAX_CLEANUP_ATTEMPTS_AT_ONCE = 10;
+
+    private final JetInstance jet;
+    private final IList<Tuple2<String, Long>> completedObservables;
+    private final long expirationTime;
+    private final int maxSize;
+    private final LongSupplier timeSource;
+
+    public ObservableRepository(JetInstance jet, JetConfig config) {
+        this(jet, config, System::currentTimeMillis);
+    }
+
+    ObservableRepository(JetInstance jet, JetConfig config, LongSupplier timeSource) {
+        this.jet = jet;
+        this.completedObservables = jet.getList(COMPLETED_OBSERVABLES_LIST_NAME);
+        this.expirationTime = getExpirationTime(config);
+        this.maxSize = getMaxSize(config);
+        this.timeSource = timeSource;
+    }
+
+    public static UUID initObservable(
+            String observable,
+            Consumer<ObservableBatch> onNewMessage,
+            Consumer<Long> onSequenceNo,
+            JetInstance jet
+    ) {
+        ITopic<ObservableBatch> topic = getTopic(jet.getHazelcastInstance(), observable);
+        return topic.addMessageListener(new TopicListener(onNewMessage, onSequenceNo));
+    }
+
+    public static void destroyObservable(
+            String observable,
+            UUID registrationId,
+            JetInstance jet
+    ) {
+        ITopic<ObservableBatch> topic = getTopic(jet.getHazelcastInstance(), observable);
+        topic.removeMessageListener(registrationId);
```

Comment:
why do we need to remove the listener before destroying? also I wonder if this method can fail.

Reply:
Not sure if we need to remove the listener, but what do we gain from not doing it? In `ReliableTopicProxy` I see a field called `runnersMap` for example, which isn't cleaned up on `destroy` so I would leave the call to `removeMessageListener`, unless I'm missing something and there would be value in not doing it.

Reply:
Seems like I will end up not removing the listener before destroying of my own accord...